### PR TITLE
Add Glue UDB to Catalog Service

### PR DESCRIPTION
**Pull request by HelloHorizon**

### CODE REVIEW

**Comment by HelloHorizon** [comment]
@gpang @yuzhu please check it

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+      }
+    }
+    mUdbContext.getFileSystem().mount(tableUri, ufsUri, option.build());
+
+    LOG.info("mounted table {} location {} to Alluxio location {} with mountOption {}",
+        tableName, ufsUri, tableUri, option.build());
+    return tableUri.getPath();
+  }
+
+  private PathTranslator mountAlluxioPaths(Table table, List<Partition> partitions)
```
Comment (yuzhu) [inline table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java:243]: similarly , can this be refactored?

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+          + tableName + ". error: " + e.getMessage(), e);
+    }
+  }
+
+  /**
+   * Get Glue Client.
+   *
+   * @return async glue client
+   */
+  public AWSGlueAsync getClient() {
```
Comment (yuzhu) [inline table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java:364]: one thing i noticed when doing the Hive version of this, is that the client sometimes does not handle disconnection well. Have you tested the case where the connection is not always stable?

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueTable.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Glue table implementation.
+ */
+public class GlueTable implements UdbTable {
```
Comment (yuzhu) [inline table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueTable.java:41]: what is the difference between this class and the HiveTable class? can we refactor to reduce duplicated code?

File: docs/en/operation/Metrics-System.md [hunk 4]
```diff
@@ -89,6 +89,7 @@ sink.csv.directory=/tmp/alluxio-metrics
 ```
 
 If Alluxio is deployed in a cluster, this file needs to be distributed to all the nodes.
+Restart the Alluxio servers to active new configuration changes.
```
[the remark this answers is not in the document]
Comment (HelloHorizon) [reply docs/en/operation/Metrics-System.md:92]: It should not be related, I will delete it later.

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+  private static AWSCredentialsProvider getAWSCredentialsProvider(UdbConfiguration config) {
+    //TODO(shouwei): add compelete authentication method for glue udb
+    if (!config.get(Property.AWS_GLUE_ACCESS_KEY).isEmpty()
+        && !config.get(Property.AWS_GLUE_SECRET_KEY).isEmpty()) {
+      return new AWSStaticCredentialsProvider(
+        new BasicAWSCredentials(
+          config.get(Property.AWS_GLUE_ACCESS_KEY),
+          config.get(Property.AWS_GLUE_SECRET_KEY)));
+    }
+    return DefaultAWSCredentialsProviderChain.getInstance();
```
[the remark this answers is not in the document]
Comment (HelloHorizon) [reply table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java:173]: The default provider chain fails all the time since they are null.

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+        tablesResult.getTableList().forEach(table -> tableNames.add(table.getName()));
+        nextToken = tablesResult.getNextToken();
+      } while (nextToken != null);
+      return tableNames;
+    } catch (EntityNotFoundException e) {
+      throw new IOException("Failed to get glue tables: " + e.getMessage(), e);
+    }
+  }
+
+  private String mountAlluxioPath(String tableName, AlluxioURI ufsUri, AlluxioURI tableUri)
```
[the remark this answers is not in the document]
Comment (HelloHorizon) [reply table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java:207]: I agree. The redundant code is also my concern, I will refactor it. It may a good idea to put these methods into udbUtil instead.

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueTable.java (hunk 3 above)
Comment (HelloHorizon) [reply table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueTable.java:41]: There is no difference for the current version, should we just use the hive layout instead?

File: table/server/common/src/main/java/alluxio/table/common/layout/GlueLayout.java [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * Glue table layout implementation.
+ * Glue use same proto as hive.
+ */
+public class GlueLayout implements Layout {
```
[the remark this answers is not in the document]
Comment (HelloHorizon) [reply table/server/common/src/main/java/alluxio/table/common/layout/GlueLayout.java:39]: There is no difference for the current version, should we just use the hive layout instead?

**Comment by HelloHorizon** [comment]
jenkins, test this please

**Comment by HelloHorizon** [comment]
jenkins, test this please

File: table/server/common/pom.xml [hunk 8]
```diff
@@ -27,6 +27,7 @@
     <!-- run properly from sub-project directories -->
     <build.path>${project.parent.parent.parent.basedir}/build</build.path>
     <failIfNoTests>false</failIfNoTests>
+    <hive.version>2.2.0</hive.version>
```
Comment (gpang) [inline table/server/common/pom.xml:30]: Is this parameter used anywhere?

File: table/server/underdb/glue/pom.xml [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+        <dependency>
+            <groupId>com.amazonaws</groupId>
+            <artifactId>aws-java-sdk-core</artifactId>
+            <version>${aws.java.jdk.version}</version>
+        </dependency>
+
+        <dependency>
+            <groupId>org.apache.hive</groupId>
+            <artifactId>hive-metastore</artifactId>
```
Comment (gpang) [inline table/server/underdb/glue/pom.xml:53]: does the glue udb connect to a hive metastore?

File: table/server/common/src/main/java/alluxio/table/common/BaseConfiguration.java [hunk 10]
```diff
@@ -55,6 +57,23 @@ public String get(T property) {
     return value;
   }
 
+  /**
+   * Return the int value of this property , or the default value if the property is not defined.
+   *
+   * @param udbProperty the udb property to get the int value
+   * @return the int value of udb property
+   */
+  public int getInt(T udbProperty) {
```
Comment (gpang) [inline table/server/common/src/main/java/alluxio/table/common/BaseConfiguration.java:66]: I think `udbProperty` should just be `property`

File: table/server/common/src/main/java/alluxio/table/common/BaseConfiguration.java [hunk 11]
```diff
@@ -55,6 +57,23 @@ public String get(T property) {
     return value;
   }
 
+  /**
+   * Return the int value of this property , or the default value if the property is not defined.
+   *
+   * @param udbProperty the udb property to get the int value
+   * @return the int value of udb property
```
Comment (gpang) [inline table/server/common/src/main/java/alluxio/table/common/BaseConfiguration.java:64]: "udb property" should just be "property"

File: table/server/common/src/main/java/alluxio/table/common/udb/UdbUtil.java [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Udb related utils.
+ */
+public class UdbUtil {
```
Comment (gpang) [inline table/server/common/src/main/java/alluxio/table/common/udb/UdbUtil.java:31]: I think we typically name these types of class with "Utils". So, it should be `UdbUtils`

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java [hunk 13]
```diff
@@ -0,0 +1,389 @@
...
+   * @param udbContext the db context
+   * @param configuration the configuration
+   * @return the new instance
+   */
+  public static GlueDatabase create(UdbContext udbContext, UdbConfiguration configuration) {
+    String glueDbName = udbContext.getUdbDbName();
+    if (glueDbName == null || glueDbName.isEmpty()) {
+      throw new IllegalArgumentException(
+          "Glue database name cannot be empty: " + glueDbName);
+    } else if (configuration.get(Property.GLUE_REGION) == null) {
```
Comment (gpang) [inline table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java:99]: is it ok if it is an empty string?

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java [hunk 14]
```diff
@@ -0,0 +1,389 @@
...
+  }
+
+  @Override
+  public UdbContext getUdbContext() {
+    return mUdbContext;
+  }
+
+  @Override
+  public DatabaseInfo getDatabaseInfo() throws IOException {
+    LOG.info("Getting Glue database information from database: " + mGlueDbName + ".");
```
Comment (gpang) [inline table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java:113]: Is this log message necessary? Should it be debug level?

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java [hunk 15]
```diff
@@ -0,0 +1,389 @@
...
+          .withCatalogId(mGlueConfiguration.get(Property.CATALOG_ID))
+          .withName(mGlueDbName);
+      GetDatabaseResult dbResult = mGlueClient.getDatabase(dbRequest);
+      Database glueDatabase = dbResult.getDatabase();
+      String glueDbLocation = glueDatabase.getLocationUri();
+      String glueDbDescription = glueDatabase.getDescription();
+      Map<String, String> glueParameters = new HashMap<>();
+      // Returned parameter can be null while Alluxio require non-null hash map for parameters
+      if (glueDatabase.getParameters() != null) {
+        glueParameters = glueDatabase.getParameters();
```
Comment (gpang) [inline table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java:125]: instead of taking the glue map, can we instead add all those parameters into `glueParameters`?

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java [hunk 16]
```diff
@@ -0,0 +1,389 @@
...
+      }
+      return new DatabaseInfo(
+          glueDbLocation,
+          mOwnerName,
+          mOwnerType,
+          glueDbDescription,
+          glueParameters);
+    } catch (EntityNotFoundException e) {
+      throw new IOException("Cannot find glue database" + mGlueDbName
```
Comment (gpang) [inline table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java:134]: This needs a space after "database" and  space before the `e.getMessage()`. Also, can we add the catalog id to the message?

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java [hunk 17]
```diff
@@ -0,0 +1,389 @@
...
+  }
+
+  /**
+   * This method allows user to test udb glue client with remote glue server.
+   *
+   * @param config udbconfiguration
+   * @return glue async client
+   */
+  @VisibleForTesting
+  protected static AWSGlueAsync createAsyncGlueClient(UdbConfiguration config) {
```
Comment (gpang) [inline table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java:146]: which test uses this?

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java (hunk 1 above)
Comment (gpang) [reply table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java:215]: Wasn't this already moved to a Utils class?

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java [hunk 18]
```diff
@@ -0,0 +1,389 @@
...
+                .withCatalogId(mGlueConfiguration.get(Property.CATALOG_ID))
+                .withDatabaseName(mGlueDbName)
+                .withNextToken(nextToken);
+        GetTablesResult tablesResult = mGlueClient.getTables(tablesRequest);
+        tablesResult.getTableList().forEach(table -> tableNames.add(table.getName()));
+        nextToken = tablesResult.getNextToken();
+      } while (nextToken != null);
+      return tableNames;
+    } catch (EntityNotFoundException e) {
+      throw new IOException("Failed to get glue tables: " + e.getMessage(), e);
```
Comment (gpang) [inline table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java:210]: can you also add catalog id and database name into the message?

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java [hunk 19]
```diff
@@ -0,0 +1,389 @@
...
+          // Glue does not provide FieldSchema from API directly
+          // Get FieldSchema from partition keys
+          GlueUtils.toProto(table.getPartitionKeys()),
+          udbPartitions,
+          layout,
+          table);
+    } catch (EntityNotFoundException e) {
+      throw new NotFoundException("Table " + tableName + " does not exist.", e);
+    } catch (ValidationException e) {
+      e.printStackTrace();
```
Comment (gpang) [inline table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java:354]: is this necessary?

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java [hunk 20]
```diff
@@ -0,0 +1,389 @@
...
+          GlueUtils.toProtoSchema(table.getStorageDescriptor().getColumns()),
+          columnStatisticsData,
+          // Glue does not provide FieldSchema from API directly
+          // Get FieldSchema from partition keys
+          GlueUtils.toProto(table.getPartitionKeys()),
+          udbPartitions,
+          layout,
+          table);
+    } catch (EntityNotFoundException e) {
+      throw new NotFoundException("Table " + tableName + " does not exist.", e);
```
Comment (gpang) [inline table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java:352]: For this exception and others in this method, can you add catalog id and database name?

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java [hunk 21]
```diff
@@ -0,0 +1,389 @@
...
+          new GetPartitionsRequest()
+              .withCatalogId(mGlueConfiguration.get(Property.CATALOG_ID))
+              .withDatabaseName(mGlueDbName)
+              .withTableName(tableName);
+      if (glueClient.getPartitions(getPartitionsRequest).getPartitions() != null) {
+        partitions = glueClient.getPartitions(getPartitionsRequest).getPartitions();
+      }
+      return partitions;
+    } catch (AWSGlueException e) {
+      throw new IOException("WARNING: Cannot get partition information for table: "
```
Comment (gpang) [inline table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java:376]: You can remove "WARNING", and also add the catalog id and the database name.

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/Property.java [hunk 22]
```diff
@@ -0,0 +1,204 @@
...
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * This contains all the properties for Glue UDB.
+ */
+public class Property extends UdbProperty {
+  private static final Logger LOG = LoggerFactory.getLogger(Property.class);
+  /** A map from default property key's string name to the key. */
+  private static final Map<String, Property> DEFAULT_KEYS_MAP = new ConcurrentHashMap<>();
```
Comment (gpang) [inline table/server/underdb/glue/src/main/java/alluxio/table/under/glue/Property.java:31]: why is this necessary?

File: table/server/underdb/glue/src/test/java/alluxio/table/under/glue/GlueDatabaseTest.java [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+
+  @Rule
+  public ExpectedException mExpection = ExpectedException.none();
+
+  private UdbContext mUdbContext;
+  private UdbConfiguration mUdbConfiguration;
+
+  @Before
+  public void before() {
+    CONF.put("aws.region", "us-east-1");
```
Comment (gpang) [inline table/server/underdb/glue/src/test/java/alluxio/table/under/glue/GlueDatabaseTest.java:41]: can we just create a local map to pass into `UdbConfiguration(conf)` on the next line?

File: table/server/underdb/glue/src/test/java/alluxio/table/under/glue/RemoteGlueTest.java [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+  private UdbConfiguration mUdbConfiguration;
+  private GlueDatabase mGlueDatabase;
+  private AWSGlueAsync mGlueClient;
+
+  private static final String AWS_ACCESS_KEY_ID = "<PUT_YOUR_ACCESS_KEY_ID_HERE>";
+  private static final String AWS_SECRET_KEY = "<PUT_YOUR_SECRET_KEY_HERE>";
+  private static final String CATALOG_ID = "<PUT_YOUR_CATALOG_ID_HERE>";
+  private static final String AWS_REGION = "<PUT_GLUE_REGION_HERE>";
+
+  @Ignore
```
Comment (gpang) [inline table/server/underdb/glue/src/test/java/alluxio/table/under/glue/RemoteGlueTest.java:44]: instead of ignoring each method, can you ignore the entire class?

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueUtils.java [hunk 25]
```diff
@@ -0,0 +1,161 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Glue utils.
+ */
+public class GlueUtils {
```
Comment (gpang) [inline table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueUtils.java:36]: For all of these conversions, does it handle when fields don't exist or are `null`? We have hit several issues with the hive one because of these scenarios.

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/Property.java [hunk 26]
```diff
@@ -0,0 +1,204 @@
...
+  }
+
+  /**
+   * Corresponding configurations of GLUE configurations.
+   */
+  public static final class Name {
+    // AWS Glue related properties
+    public static final String MAX_GLUE_CONNECTION = "max.connection";
+    public static final String GLUE_REGION = "aws.region";
+    public static final String CATALOG_ID = "aws.catalog.id";
```
Comment (gpang) [inline table/server/underdb/glue/src/main/java/alluxio/table/under/glue/Property.java:200]: Did you figure out if this is required or not?

File: table/server/common/pom.xml (hunk 8 above)
Comment (HelloHorizon) [reply table/server/common/pom.xml:30]: I will delete it.

File: table/server/common/src/main/java/alluxio/table/common/BaseConfiguration.java (hunk 11 above)
Comment (HelloHorizon) [reply table/server/common/src/main/java/alluxio/table/common/BaseConfiguration.java:64]: Fixed it.

File: table/server/common/src/main/java/alluxio/table/common/udb/UdbUtil.java (hunk 12 above)
Comment (HelloHorizon) [reply table/server/common/src/main/java/alluxio/table/common/udb/UdbUtil.java:31]: Fixed it.

File: table/server/common/src/main/java/alluxio/table/common/BaseConfiguration.java (hunk 10 above)
Comment (HelloHorizon) [reply table/server/common/src/main/java/alluxio/table/common/BaseConfiguration.java:66]: Fixed it.

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java (hunk 14 above)
Comment (HelloHorizon) [reply table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java:113]: Fixed it.

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java (hunk 17 above)
Comment (HelloHorizon) [reply table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java:146]: Deleted.

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java (hunk 18 above)
Comment (HelloHorizon) [reply table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java:210]: The mountAlluxioPath is already moved to UdbUtils. I can add the catalog id and database name into the message.

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java (hunk 19 above)
Comment (HelloHorizon) [reply table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java:354]: Removed.

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java (hunk 1 above)
Comment (HelloHorizon) [reply table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java:215]: Since mountAlluxioPaths required dependency on hive and glue, which is a different method for hive udb and glue udb. So I move mountAlluxioPath to Utils class, and keep mountAlluxioPaths in hive udb and glue udb.

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueUtils.java (hunk 25 above)
Comment (HelloHorizon) [reply table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueUtils.java:36]: I have fixed some of the `null` filed issues in `glueUtils`, I will add more UT for these scenarios.

File: table/server/underdb/glue/src/test/java/alluxio/table/under/glue/GlueDatabaseTest.java (hunk 23 above)
Comment (HelloHorizon) [reply table/server/underdb/glue/src/test/java/alluxio/table/under/glue/GlueDatabaseTest.java:41]: Changed.

File: table/server/underdb/glue/src/test/java/alluxio/table/under/glue/RemoteGlueTest.java (hunk 24 above)
Comment (HelloHorizon) [reply table/server/underdb/glue/src/test/java/alluxio/table/under/glue/RemoteGlueTest.java]: Changed.

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java (hunk 15 above)
Comment (HelloHorizon) [reply table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java]: Changed.

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java (hunk 16 above)
Comment (HelloHorizon) [reply table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java]: Changed.

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java (hunk 20 above)
Comment (HelloHorizon) [reply table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java:352]: Added.

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java (hunk 21 above)
Comment (HelloHorizon) [reply table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java]: Fixed.

**Comment by HelloHorizon** [comment]
jenkins, test this please

**Comment by HelloHorizon** [comment]
jenkins, test this please

**Comment by HelloHorizon** [comment]
jenkins, test this please

**Comment by HelloHorizon** [comment]
alluxio-bot, test this please

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java (hunk 13 above)
Comment (gpang) [reply table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java:99]: Is empty string region allowed?

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java (hunk 18 above)
Comment (gpang) [reply table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java:210]: Can you add the db name and catalog id?

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java (hunk 20 above)
Comment (gpang) [reply table/server/underdb/glue/src/main/java/alluxio/table/under/glue/GlueDatabase.java:352]: I don't see the updated exception messages which include the db name and catalog id.

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/Property.java [hunk 27]
```diff
@@ -0,0 +1,204 @@
...
+  }
+
+  /**
+   * Registers the given UDB alluxio.table.under.glue.Property to the global map.
+   *
+   * @param Property the udb property
+   * @return whether the udb property is successfully registered
+   */
+  @VisibleForTesting
+  public static boolean register(Property Property) {
```
[the remark this answers is not in the document]
Comment (HelloHorizon) [reply table/server/underdb/glue/src/main/java/alluxio/table/under/glue/Property.java:118]: The register is used to build a new property, and there's no use for unregister for now.

File: table/server/underdb/glue/src/test/java/alluxio/table/under/glue/GlueDatabaseTest.java [hunk 28]
```diff
@@ -0,0 +1,68 @@
...
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class GlueDatabaseTest {
+
+  private static final String DB_NAME = "test";
+  private static final Map<String, String> CONF = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (HelloHorizon) [reply table/server/underdb/glue/src/test/java/alluxio/table/under/glue/GlueDatabaseTest.java:31]: The property is needed since glue need some extra properties compare to default udb.
I will change it to local map.

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/Property.java (hunk 26 above)
Comment (HelloHorizon) [reply table/server/underdb/glue/src/main/java/alluxio/table/under/glue/Property.java:200]: Added to TODO.

**Comment by HelloHorizon** [comment]
alluxio-bot, test this please

File: table/server/underdb/glue/src/main/java/alluxio/table/under/glue/Property.java (hunk 26 above)
Comment (gpang) [reply table/server/underdb/glue/src/main/java/alluxio/table/under/glue/Property.java:200]: I don't see the TODO comment?

**Comment by gpang** [comment]
alluxio-bot, merge this please
